resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1

```diff
@@ -0,0 +1,547 @@
+        $inputObject = @()
+
+        # Skip the header lines and process each line
+        foreach ($line in $lines)
+        {
+            # Split the line into columns
+            $columns = $line -split '\s{2,}'
+
+            # Create a custom object for each line
+            $customObject = [PSCustomObject]@{
+                PackageId = $columns[0]
+                Version   = $columns[1]
+                Commands  = $columns[2]
+            }
+
+            # Add the custom object to the array
+            $inputObject += $customObject
+        }
+
+        return $inputObject
+    }
+}
+
+function Get-InstalledDotNetToolPackages
+{
+    [CmdletBinding()]
+    param (
+        [string] $PackageId,
+        [string] $Version,
+        [bool]   $PreRelease,
+        [Parameter(Mandatory = $false)]
+        [ValidateScript({
+                if (-Not ($_ | Test-Path -PathType Container) )
+                {
+                    throw "Directory does not exist" 
+                }
+                return $true
+            })]
+        [string] $ToolPath,
+        [bool]   $Exist
+    )
+
+    $resultSet = [System.Collections.Generic.List[DotNetToolPackage]]::new()
+    $listCommand = "tool list --global"
+    $installDir = Join-Path -Path $env:USERPROFILE '.dotnet' 'tools'
+
+    if ($PSBoundParameters.ContainsKey('ToolPath'))
+    {
+        $listCommand = "tool list --tool-path $ToolPath"
+        $installDir = $ToolPath
+    }
+
+    $result = Invoke-DotNet -Command $listCommand
+    $packages = Convert-DotNetToolOutput -Output $result
+
+    if ($null -eq $packages)
+    {
+        Write-Debug -Message "No packages found."
+        return
+    }
+
+    if (-not [string]::IsNullOrEmpty($PackageId))
+    {
+        $packages = $packages | Where-Object { $_.PackageId -eq $PackageId }
+    }
+
+    foreach ($package in $packages)
+    {
+        # flags to determine the existence of the package
+        $preFlag = $false
+        $preReleasePackage = $package.Version -Split "-"
+        if ($preReleasePackage.Count -gt 1)
+        {
+            # set the pre-release flag to true to build the object
+            $preFlag = $true
+        }
+
+        $resultSet.Add([DotNetToolPackage]::new(
+                $package.PackageId, $package.Version, $package.Commands, $preFlag, $installDir, $true
+            ))
+    }
+
+    return $resultSet
+}
+
+function Get-SemVer($version)
+{
+    $version -match "^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?(\-(?<pre>[0-9A-Za-z\-\.]+))?(\+(?<build>[0-9A-Za-z\-\.]+))?$" | Out-Null
+    $major = [int]$matches['major']
+    $minor = [int]$matches['minor']
+    $patch = [int]$matches['patch']
+    
+    if ($null -eq $matches['pre']) { $pre = @() }
+    else { $pre = $matches['pre'].Split(".") }
+
+    $revision = 0
+    if ($pre.Count -gt 0)
+    {
+        $revision = Get-HighestRevision -InputArray $pre
+    }
+
+    return [version]$version = "$major.$minor.$patch.$revision" 
+}
```

Comment:
Can consider adding some basic unit tests for this function to verify the behavior